File: src/main/java/net/atos/client/zgw/zrc/model/RolVestiging.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.zrc.model;

import java.net.URI;
import java.util.Objects;

import org.apache.commons.lang3.StringUtils;

import nl.info.client.zgw.ztc.model.generated.RolType;

public class RolVestiging extends Rol<Vestiging> {

    public RolVestiging() {
    }

    public RolVestiging(
            final URI zaak,
            final RolType roltype,
            final String roltoelichting,
            final Vestiging betrokkeneIdentificatie
    ) {
        super(zaak, roltype, BetrokkeneType.VESTIGING, betrokkeneIdentificatie, roltoelichting);
    }

    @Override
    protected boolean equalBetrokkeneIdentificatie(final Vestiging identificatie) {
        final Vestiging betrokkeneIdentificatie = getBetrokkeneIdentificatie();
        if (betrokkeneIdentificatie == identificatie) {
            return true;
        }
        if (identificatie == null) {
            return false;
        }
        return Objects.equals(betrokkeneIdentificatie.getVestigingsNummer(), identificatie.getVestigingsNummer());
    }

    @Override
    public String getNaam() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        final String namen = getBetrokkeneIdentificatie().getHandelsnaam() != null ? String.join("; ", getBetrokkeneIdentificatie()
                .getHandelsnaam()) : null;
        return StringUtils.isNotEmpty(namen) ? namen : getIdentificatienummer();
    }

    @Override
    public String getIdentificatienummer() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        return getBetrokkeneIdentificatie().getVestigingsNummer();
    }

    @Override
    protected int hashCodeBetrokkeneIdentificatie() {
        return Objects.hash(getBetrokkeneIdentificatie().getVestigingsNummer());
    }
}

File: src/main/kotlin/nl/info/client/zgw/zrc/model/RolVestiging.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.zrc.model

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum
import nl.info.client.zgw.zrc.model.generated.VestigingIdentificatie
import nl.info.client.zgw.ztc.model.generated.RolType
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.util.Objects

/**
 * Manually copied from [nl.info.client.zgw.zrc.model.generated.RolVestiging] and modified to allow for
 * polymorphism using a generic base [Rol] class.
 * Ideally we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 *
 * In ZAC, we only use the vestiging role for retrieving existing roles, but not / no longer for storing new roles.
 * For storing new vestiging-type roles, please use [nl.info.client.zgw.zrc.model.generated.RolNietNatuurlijkPersoon] instead for
 * vestigingen.
 * For details see: [Add 'vestigingsnummer' to NNP, deprecate 'vestiging'](https://github.com/open-zaak/open-zaak/issues/1935).
 */
class RolVestiging : Rol<VestigingIdentificatie> {
    constructor() : super()

    constructor(
        zaak: URI,
        roltype: RolType,
        roltoelichting: String,
        betrokkeneIdentificatie: VestigingIdentificatie?
    ) : super(zaak, roltype, BetrokkeneTypeEnum.VESTIGING, betrokkeneIdentificatie, roltoelichting)

    @Suppress("ReturnCount")
    override fun equalBetrokkeneIdentificatie(identificatie: VestigingIdentificatie?): Boolean {
        val ownIdentificatie = betrokkeneIdentificatie
        if (ownIdentificatie === identificatie) {
            return true
        }
        if (identificatie == null) {
            return false
        }
        return Objects.equals(ownIdentificatie?.vestigingsNummer, identificatie.vestigingsNummer)
    }

    override val naam: String?
        get() {
            val identificatie = betrokkeneIdentificatie ?: return null
            val namen = identificatie.handelsnaam?.joinToString("; ")
            return if (StringUtils.isNotEmpty(namen)) namen else identificatienummer
        }

    override val identificatienummer: String?
        get() = betrokkeneIdentificatie?.vestigingsNummer

    override fun hashCodeBetrokkeneIdentificatie(): Int = Objects.hash(betrokkeneIdentificatie?.vestigingsNummer)
}
